Inductor logging + analysis of torch.profile

exclamaforte opened the pull request:
Prereqs:
 - https://github.com/pytorch/pytorch/pull/152708

Features:
1. Adds inductor's estimate of flops and bandwidth to the json trace events that perfetto uses.
1. Only use the tflops estimation from triton if we don't have the info from the datasheet because Triton's estimates are inaccurate. I have a backlog item to fix triton flops estimation upstream. New `DeviceInfo` class, and new function `get_device_tflops`.
1. New helpers `countable_fx` and `count_flops_fx` helps get the flops of an `fx.Node`.
1. Extends Triton `torch.profiler` logging to `DebugAutotuner`.
1. New script `profile_analysis.py`: `--augment_trace` adds perf estimates to any perfetto json trace, `--analyze` creates a summary table of these perf estimates, and `--diff` will compare two traces side by side:
```python
Device(NVIDIA H100, 0):
 Kernel Name                              | resnet Kernel Count | resnet FLOPS       | resnet bw gbps        | resnet Dur (ms)    | resnet Achieved FLOPS % | resnet Achieved Bandwidth % | newresnet Kernel Count | newresnet FLOPS    | newresnet bw gbps     | newresnet Dur (ms) | newresnet Achieved FLOPS % | newresnet Achieved Bandwidth % 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 triton_poi_fused__native_batch_norm_legi | 24                  | 0                  | 0.11395268248131513   | 2.5919166666666666 | 0                       | 0.003401572611382541        | 24                     | 0                  | 0.11395268248131513   | 2.5919166666666666 | 0                          | 0.003401572611382541           
 sm90_xmma_fprop_implicit_gemm_f32f32_tf3 | 142                 | 16932673552.422373 | 0.2585007824198784    | 12.441619718309857 | 0.08683422334575583     | 0.007716441266265022        | 142                    | 16932673552.422373 | 0.2585007824198784    | 12.441619718309857 | 0.08683422334575583        | 0.007716441266265022           
 triton_red_fused__native_batch_norm_legi | 39                  | 0                  | 0.13990024992108846   | 5.752589743589743  | 0                       | 0.004176126863316074        | 39                     | 0                  | 0.13990024992108846   | 5.752589743589743  | 0                          | 0.004176126863316074           
 triton_poi_fused__native_batch_norm_legi | 25                  | 0                  | 0.31824055917536503   | 2.5291999999999994 | 0                       | 0.009499718184339253        | 25                     | 0                  | 0.31824055917536503   | 2.5291999999999994 | 0                          | 0.009499718184339253           
 void cutlass::Kernel2<cutlass_80_tensoro | 98                  | 16211056473.596165 | 0.42972434051025826   | 7.130408163265306  | 0.08313362294151874     | 0.012827592254037562        | 98                     | 16211056473.596165 | 0.42972434051025826   | 7.130408163265306  | 0.08313362294151874        | 0.012827592254037562           
 triton_red_fused__native_batch_norm_legi | 73                  | 0                  | 0.3225381327611705    | 9.987068493150682  | 0                       | 0.009628003963020014        | 73                     | 0                  | 0.3225381327611705    | 9.987068493150682  | 0                          | 0.009628003963020014           
 triton_poi_fused__native_batch_norm_legi | 15                  | 0                  | 1.4491211346487216    | 4.439333333333333  | 0                       | 0.043257347302946926        | 15                     | 0                  | 1.4491211346487216    | 4.439333333333333  | 0                          | 0.043257347302946926           
 void cutlass::Kernel2<cutlass_80_tensoro | 186                 | 14501701145.337954 | 0.2667131401910989    | 7.873865591397849  | 0.07436769818122027     | 0.007961586274361157        | 186                    | 14501701145.337954 | 0.2667131401910989    | 7.873865591397849  | 0.07436769818122027        | 0.007961586274361157           
 triton_poi_fused__native_batch_norm_legi | 33                  | 0                  | 1.4924556538193923    | 4.3101515151515155 | 0                       | 0.044550915039384846        | 33                     | 0                  | 1.4924556538193923    | 4.3101515151515155 | 0                          | 0.044550915039384846           
 triton_red_fused__native_batch_norm_legi | 29                  | 0                  | 0.25562590522631107   | 6.296275862068965  | 0                       | 0.007630624036606301        | 29                     | 0                  | 0.25562590522631107   | 6.296275862068965  | 0                          | 0.007630624036606301           
 triton_poi_fused__native_batch_norm_legi | 13                  | 0                  | 0.5870562174192726    | 2.7397692307692307 | 0                       | 0.01752406619162008         | 13                     | 0                  | 0.5870562174192726    | 2.7397692307692307 | 0                          | 0.01752406619162008            
 triton_poi_fused__native_batch_norm_legi | 34                  | 0                  | 0.41409928846284      | 2.853588235294117  | 0                       | 0.012361172789935523        | 34                     | 0                  | 0.41409928846284      | 2.853588235294117  | 0                          | 0.012361172789935523           
 triton_per_fused__native_batch_norm_legi | 34                  | 0                  | 0.11705315007018151   | 3.460647058823529  | 0                       | 0.0034941238826919864       | 34                     | 0                  | 0.11705315007018151   | 3.460647058823529  | 0                          | 0.0034941238826919864          
 triton_poi_fused__native_batch_norm_legi | 16                  | 0                  | 0.17207853197124584   | 2.3459375000000002 | 0                       | 0.005136672596156592        | 16                     | 0                  | 0.17207853197124584   | 2.3459375000000002 | 0                          | 0.005136672596156592           
 triton_per_fused__native_batch_norm_legi | 30                  | 0                  | 0.2639714322022256    | 6.131199999999999  | 0                       | 0.007879744244842555        | 30                     | 0                  | 0.2639714322022256    | 6.131199999999999  | 0                          | 0.007879744244842555           
 sm90_xmma_fprop_implicit_gemm_f32f32_tf3 | 100                 | 11875430356.891787 | 0.19494470869421385   | 16.36534           | 0.06089964285585531     | 0.005819245035648175        | 100                    | 11875430356.891787 | 0.19494470869421385   | 16.36534           | 0.06089964285585531        | 0.005819245035648175           
 triton_poi_fused__native_batch_norm_legi | 8                   | 0                  | 0.9854096626224687    | 3.2757500000000004 | 0                       | 0.029415213809625928        | 8                      | 0                  | 0.9854096626224687    | 3.2757500000000004 | 0                          | 0.029415213809625928           
 void cublasLt::splitKreduce_kernel<32, 1 | 56                  | 34377923395.147064 | 0.8310300045762317    | 3.4199999999999986 | 0.17629704305203628     | 0.024806865808245714        | 56                     | 34377923395.147064 | 0.8310300045762317    | 3.4199999999999986 | 0.17629704305203628        | 0.024806865808245714           
 triton_poi_fused__native_batch_norm_legi | 23                  | 0                  | 0.9944002965861103    | 3.2431304347826084 | 0                       | 0.02968359094286896         | 23                     | 0                  | 0.9944002965861103    | 3.2431304347826084 | 0                          | 0.02968359094286896            
 triton_per_fused__native_batch_norm_legi | 10                  | 0                  | 0.1826801058931057    | 4.428800000000001  | 0                       | 0.00545313748934644         | 10                     | 0                  | 0.1826801058931057    | 4.428800000000001  | 0                          | 0.00545313748934644            
 triton_poi_fused__native_batch_norm_legi | 10                  | 0                  | 0.3168973585366449    | 2.5471999999999997 | 0                       | 0.009459622642884923        | 10                     | 0                  | 0.3168973585366449    | 2.5471999999999997 | 0                          | 0.009459622642884923           
 triton_poi_fused__native_batch_norm_legi | 34                  | 0                  | 1.1463614897015777    | 4.124323529411764  | 0                       | 0.03421974596124114         | 34                     | 0                  | 1.1463614897015777    | 4.124323529411764  | 0                          | 0.03421974596124114            
 void cask_plugin_cudnn::xmma_cudnn::init | 44                  | 44045510816.64277  | 2.0661232850348643    | 3.6887499999999993 | 0.22587441444432194     | 0.06167532194133924         | 44                     | 44045510816.64277  | 2.0661232850348643    | 3.6887499999999993 | 0.22587441444432194        | 0.06167532194133924            
 sm90_xmma_fprop_implicit_gemm_f32f32_tf3 | 95                  | 7876855400.165316  | 0.4694941555946739    | 18.224315789473682 | 0.04039413025725802     | 0.014014750913273854        | 95                     | 7876855400.165316  | 0.4694941555946739    | 18.224315789473682 | 0.04039413025725802        | 0.014014750913273854           
 triton_per_fused__native_batch_norm_legi | 41                  | 0                  | 0.06825669875995298   | 3.0384146341463416 | 0                       | 0.002037513395819492        | 41                     | 0                  | 0.06825669875995298   | 3.0384146341463416 | 0                          | 0.002037513395819492           
 triton_poi_fused__native_batch_norm_legi | 23                  | 0                  | 0.08808154712430301   | 2.3275652173913044 | 0                       | 0.0026292999141582997       | 23                     | 0                  | 0.08808154712430301   | 2.3275652173913044 | 0                          | 0.0026292999141582997          
 triton_per_fused__native_batch_norm_legi | 40                  | 0                  | 0.18179321034952417   | 4.556825           | 0                       | 0.005426662995508183        | 40                     | 0                  | 0.18179321034952417   | 4.556825           | 0                          | 0.005426662995508183           
 triton_poi_fused__native_batch_norm_legi | 15                  | 0                  | 0.5887415155454232    | 2.783866666666667  | 0                       | 0.017574373598370836        | 15                     | 0                  | 0.5887415155454232    | 2.783866666666667  | 0                          | 0.017574373598370836           
 void cutlass::Kernel2<cutlass_80_tensoro | 38                  | 14242013806.264643 | 0.256592404353939     | 7.217631578947369  | 0.0730359682372546      | 0.007659474756834           | 38                     | 14242013806.264643 | 0.256592404353939     | 7.217631578947369  | 0.0730359682372546         | 0.007659474756834              
 triton_poi_fused__native_batch_norm_legi | 21                  | 0                  | 0.5842860973430516    | 2.7779047619047623 | 0                       | 0.017441376040091088        | 21                     | 0                  | 0.5842860973430516    | 2.7779047619047623 | 0                          | 0.017441376040091088           
 triton_per_fused__native_batch_norm_legi | 16                  | 0                  | 0.11509365173486417   | 3.5959375000000002 | 0                       | 0.0034356313950705724       | 16                     | 0                  | 0.11509365173486417   | 3.5959375000000002 | 0                          | 0.0034356313950705724          
 triton_poi_fused__native_batch_norm_legi | 14                  | 0                  | 0.1704672000243914    | 2.4044285714285714 | 0                       | 0.00508857313505646         | 14                     | 0                  | 0.1704672000243914    | 2.4044285714285714 | 0                          | 0.00508857313505646            
 triton_poi_fused__native_batch_norm_legi | 58                  | 0                  | 2.307520779930795     | 8.190706896551722  | 0                       | 0.06888121731136704         | 58                     | 0                  | 2.307520779930795     | 8.190706896551722  | 0                          | 0.06888121731136704            
 triton_per_fused__native_batch_norm_legi | 29                  | 0                  | 0.037243248971881276  | 3.0277586206896556 | 0                       | 0.001111738775280038        | 29                     | 0                  | 0.037243248971881276  | 3.0277586206896556 | 0                          | 0.001111738775280038           
 triton_poi_fused__native_batch_norm_legi | 20                  | 0                  | 0.04741699795428918   | 2.2911500000000005 | 0                       | 0.0014154327747549007       | 20                     | 0                  | 0.04741699795428918   | 2.2911500000000005 | 0                          | 0.0014154327747549007          
 triton_per_fused__native_batch_norm_legi | 25                  | 0                  | 0.13357016893727824   | 3.37536            | 0                       | 0.003987169222008305        | 25                     | 0                  | 0.13357016893727824   | 3.37536            | 0                          | 0.003987169222008305           
 triton_poi_fused__native_batch_norm_legi | 13                  | 0                  | 0.3089862268300253    | 2.8111538461538457 | 0                       | 0.009223469457612694        | 13                     | 0                  | 0.3089862268300253    | 2.8111538461538457 | 0                          | 0.009223469457612694           
 triton_poi_fused__native_batch_norm_legi | 17                  | 0                  | 0.3129385387909844    | 2.673              | 0                       | 0.009341448919133863        | 17                     | 0                  | 0.3129385387909844    | 2.673              | 0                          | 0.009341448919133863           
 triton_per_fused__native_batch_norm_legi | 19                  | 0                  | 0.2215568162533158    | 3.8837368421052636 | 0                       | 0.0066136363060691275       | 19                     | 0                  | 0.2215568162533158    | 3.8837368421052636 | 0                          | 0.0066136363060691275          
 std::enable_if<!(false), void>::type int | 23                  | 504916805.19297093 | 1.0118296096314707    | 8.113913043478261  | 0.0025893169497075447   | 0.030203868944223014        | 23                     | 504916805.19297093 | 1.0118296096314707    | 8.113913043478261  | 0.0025893169497075447      | 0.030203868944223014           
 triton_poi_fused_add_copy__38            | 56                  | 0                  | 0                     | 2.132482142857143  | 0                       | 0                           | 56                     | 0                  | 0                     | 2.132482142857143  | 0                          | 0                              
 triton_poi_fused_convolution_0           | 18                  | 0                  | 0.43458610794936897   | 2.773333333333334  | 0                       | 0.012972719640279667        | 18                     | 0                  | 0.43458610794936897   | 2.773333333333334  | 0                          | 0.012972719640279667           
 triton_poi_fused_convolution_1           | 17                  | 0                  | 0.028816312469162712  | 2.6145882352941174 | 0                       | 0.0008601884319153051       | 17                     | 0                  | 0.028816312469162712  | 2.6145882352941174 | 0                          | 0.0008601884319153051          
 void convolve_common_engine_float_NHWC<f | 44                  | 8641868995.31118   | 0.024730540008465626  | 25.87327272727273  | 0.04431727689903169     | 0.0007382250748795709       | 44                     | 8641868995.31118   | 0.024730540008465626  | 25.87327272727273  | 0.04431727689903169        | 0.0007382250748795709          
 triton_per_fused__native_batch_norm_legi | 12                  | 0                  | 0.6809930918986744    | 4.82675            | 0                       | 0.020328151996975356        | 12                     | 0                  | 0.6809930918986744    | 4.82675            | 0                          | 0.020328151996975356           
 triton_per_fused__native_batch_norm_legi | 14                  | 0                  | 0.02883030597936608   | 2.6651428571428575 | 0                       | 0.0008606061486377935       | 14                     | 0                  | 0.02883030597936608   | 2.6651428571428575 | 0                          | 0.0008606061486377935          
 triton_per_fused__native_batch_norm_legi | 16                  | 0                  | 0.0014658988233201874 | 2.098              | 0                       | 4.375817383045335e-05       | 16                     | 0                  | 0.0014658988233201874 | 2.098              | 0                          | 4.375817383045335e-05          
 triton_poi_fused__native_batch_norm_legi | 13                  | 0                  | 0.9926297180284697    | 3.2367692307692306 | 0                       | 0.02963073785159611         | 13                     | 0                  | 0.9926297180284697    | 3.2367692307692306 | 0                          | 0.02963073785159611            
 triton_poi_fused__native_batch_norm_legi | 9                   | 0                  | 1.3008817095666507    | 3.0863333333333336 | 0                       | 0.03883228983781048         | 9                      | 0                  | 1.3008817095666507    | 3.0863333333333336 | 0                          | 0.03883228983781048            
 void at::native::(anonymous namespace):: | 98                  | 0                  | 0.09174335613709389   | 4.408520408163265  | 0                       | 0.0027386076458833994       | 98                     | 0                  | 0.09174335613709389   | 4.408520408163265  | 0                          | 0.0027386076458833994          
 void at::native::vectorized_elementwise_ | 7                   | 0                  | 0                     | 1.7278571428571428 | 0                       | 0                           | 7                      | 0                  | 0                     | 1.7278571428571428 | 0                          | 0                              
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

shunting314 commented:
Why matmul/convolution is missing in the diff table?

shunting314 commented on torch/_inductor/wrapper_benchmark.py:
```diff
@@ -294,6 +295,86 @@ def report() -> None:
+
+        headers = [
+            "Kernel Name",
+            f"{filename1} FLOPS",
```
Should number of calls and total latency also be relevant?

e.g., if a call only takes 1ns, even though the bw/flops is low, should not be that a big deal

exclamaforte replied:
adding

shunting314 commented on torch/_inductor/scheduler.py:
```diff
@@ -736,6 +737,50 @@ def get_buf_bytes(
+                    cls.process_kernel(op, *fake_inputs, **kern.node.kwargs)
+
+                    ret = flop_counter_mode.get_total_flops()
+                    with open("flops", "a") as file:
+                        file.write(str(kern.node) + "\nFlops: " + str(ret) + "\n")
```
Remove before merging

shunting314 commented on torch/_inductor/wrapper_benchmark.py:
```diff
@@ -294,6 +295,86 @@ def report() -> None:
+KernelNameMap = defaultdict[str, OrderedSet[KernelStats]]
+
+
+def diff_profiles(diff_path: str) -> None:
```
Maybe the diff tool should be a standalone script to compare two different profiles

exclamaforte replied:
Where do you think this should live?

shunting314 replied:
I'm fine with adding it under torch/_inductor folder directly.

shunting314 commented on torch/_inductor/wrapper_benchmark.py:
```diff
@@ -294,6 +295,86 @@ def report() -> None:
+            ):
+                name_map[event["name"]].add(
+                    KernelStats(
+                        event["args"]["kernel_flops"], event["args"]["kernel_bandwidth"]
```
Do you need to do anything else to make kernel_flops/kernel_bandwidth available in the profile?

exclamaforte replied:
There's nothing else needed, they're in the args section when you click on a slice.

shunting314 replied:
or the question is, are those args in the profile copied directly from inductor_meta? Fallback kernels don't have those

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -736,6 +737,48 @@ def get_buf_bytes(
         return buf_byte_accesses
 
+    @cache_on_self
+    def estimate_flops(self) -> int | None:
```
How do we want to handle symints ? we would need to augment caching autotuner with a sympy expression based on inputs. 

FlopCounter should work with sympy. see, colab here: https://colab.research.google.com/drive/1zjAisRrc8R6uixKsrs1DRm3lwz5MWN68#scrollTo=zbl9to6G02vQ

exclamaforte replied:
@eellison So symints ended up breaking this feature when used in max autotune, because it would generate an `inductor_meta` like:
```python
inductor_meta={'kernel_name': 'Placeholder.DESCRIPTIVE_NAME', 'backend_hash': '969E2A31A8F620FC2E1E37975F57012FF2CF1D2549D5BE17FB534BA8B4F5EF84', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'max_autotune': True, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'grid_type': 'FixedGrid', 'fixed_grid': ['_grid_0', '_grid_1', '_grid_2'], 'extra_launcher_args': ['_grid_0', '_grid_1', '_grid_2'], 'kernel_num_gb': 0.042842176, 'kernel_flop': 1536*s27*s38},
```
and would throw an error:
```
File "/tmp/tmpbx3d5iaf/4y/c4yldhzvilvh5h35ih35aud3c42bbzeuuvxspmufnlojgth7yxgy.py", line 20, in <module>
torch._inductor.exc.InductorError: NameError: name 's27' is not defined
```

So I added a line like:
```python
        resolved_flops = V.graph.sizevars.size_hints((flops,))[0]
```

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -736,6 +737,48 @@ def get_buf_bytes(
+        from torch._subclasses.fake_tensor import FakeTensorMode
+        from torch.utils.flop_counter import FlopCounterMode
+
+        op = kernel_name_to_op.get(getattr(self.node, "python_kernel_name", ""), None)
```
can you check that we preserve flops with fusion, say mm + relu?

exclamaforte replied:
added

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -457,6 +457,8 @@ def jit_lines(self):
         if config.profile_bandwidth or config.benchmark_kernel:
             num_gb = self.estimate_kernel_num_bytes() / 1e9
             inductor_meta["kernel_num_gb"] = num_gb
+            flops = self.estimate_flops()
```
do we have any tests for the Triton Templates?

eellison reviewed (commented):
Looking good, a few comments. 

General questions: 

- What is this enabled under, I see config.profile_bandwidth and config. benchmark_kernel ?

- I think not being on by default makes sense. But just to say it - if this is on by default, we need to be careful about interfering with caching by annotating the configs. But anyways, not a real issue.

- Could be nice to express flops/memory bandwidth as expression of symints. bc today they could be misleading with different shapes..

-  Ultimately we are going to want not to look at memory bandwidth or flops as % of hardware limits. so would be great to enable that as part of perfetto if possible.  

For my own understanding - how does this get into perfetto trace ? and is there anyway to augment it more outside of perfetto (like for calculating memory bw based on symint inputs). cc @davidberard98 who is knowledgeable on profiler.

shunting314 commented on torch/_inductor/wrapper_benchmark.py:
```diff
@@ -294,6 +295,86 @@ def report() -> None:
+            data = json.load(f)
+        return parse(data)
+
+    other_nm = parse_helper(PROFILE_PATH)
```
I think it's more clear to decouple profile collecting and profile comparison.

exclamaforte commented:
@eellison

> * What is this enabled under, I see config.profile_bandwidth and config. benchmark_kernel ?

The `inductor_meta` annotations will be triggered if either is set. They will only be added to torch.profile whenever `autograd_profiler._is_profiler_enabled`, meaning when the code is surrounded in a with `torch.profile` block.

> * I think not being on by default makes sense. But just to say it - if this is on by default, we need to be careful about interfering with caching by annotating the configs. But anyways, not a real issue.

You mean adding to `inductor_meta` will prevent caching?

> * Could be nice to express flops/memory bandwidth as expression of symints. bc today they could be misleading with different shapes..

Good point, I'll take a look.

> * Ultimately we are going to want not to look at memory bandwidth or flops as % of hardware limits. so would be great to enable that as part of perfetto if possible.

On it.

> For my own understanding - how does this get into perfetto trace ? and is there anyway to augment it more outside of perfetto (like for calculating memory bw based on symint inputs). cc @davidberard98 who is knowledgeable on profiler.

```python
                "kernel_bandwidth": self.inductor_meta.get("kernel_num_gb", None),
                "kernel_flops": self.inductor_meta.get("kernel_flops", None),
```
These two lines add it to the `args` field on a "slice" in the json profile. In perfetto, you can click on a "slice" and see all the values in `args`. We can add any string we want to these, so adding % of hardware or in terms of symints is no problem.

eellison commented:
The symints is not needed for this pr ! i think other things more important. but could be cool in the future.

sraikund16 commented:
Overall seems fine to me. Just wondering if we think we will want this on by default? Many users have been complaining about the size of the traces

eellison commented:
Test failures ?

eellison commented on torch/_inductor/utils.py:
```diff
@@ -2838,3 +2856,61 @@ def get_ld_library_path() -> str:
     return path
+
+
+def tabulate_2d(elements: Sequence[Sequence[T]], headers: Sequence[T]) -> str:
```
Is there a reason to not use `tabulate.tabulate` here ?

exclamaforte replied:
The CI runners don't have it. I don't think any of the other tabulate code being run has CI tests.

eellison replied:
Could we just use tabulate ? there are many other uses of tabulate in torch/_inductor/

eellison reviewed (commented):
Excited for this ! Left some comments.

cc @sraikund16 can you review as well ? At least the profiler relevant parts.

Also, today, a user will have to manually run the post processing step. It would be great if  there were an ap to register a post process fn that will be called before the profile is saved so this gets added automatically.

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -819,15 +854,13 @@ def get_estimated_runtime(self) -> float:
 
         if isinstance(self, ExternKernelSchedulerNode):
```
Looks like that will also fix the failures with this pr. 

You can reuse/factor out this:

https://github.com/pytorch/pytorch/blob/4843ce7611b5b9e85b16f1c05cd3b4959c166fce/torch/_inductor/graph.py#L632-L655

exclamaforte commented on torch/_inductor/analysis/device_info.py:
```diff
@@ -0,0 +1,136 @@
+# TODO investigate profiler support for tf32 and allow device to report correct number when it's turned on.
+_device_mapping: dict[str, DeviceInfo] = {
+    # Source: https://resources.nvidia.com/en-us-tensor-core/nvidia-tensor-core-gpu-datasheet
+    "NVIDIA H100": DeviceInfo(
```
This is based on `torch.cuda.get_device_name()`, which is stored by the profiler and available at runtime too. I'm not sure how to distinguish them, even at runtime.
Some ideas:
```
>>> torch.cuda.get_device_properties()
_CudaDeviceProperties(name='NVIDIA H100', major=9, minor=0, total_memory=97272MB, multi_processor_count=132, uuid=6efc17fa-5b7e-0452-613b-df241e45f2b8, L2_cache_size=60MB)
```
```
>>> torch.cuda.mem_get_info()
(99949740032, 101997215744)
```

exclamaforte commented on torch/_inductor/analysis/device_info.py:
```diff
@@ -0,0 +1,136 @@
+        dram_gb=80,
+    ),
+    # Source: https://resources.nvidia.com/en-us-tensor-core/nvidia-tensor-core-gpu-datasheet
+    "NVIDIA A100": DeviceInfo(
```
yeah I saw that, I'm not sure how we solve in general for bw (flops seems fine). Like on 8x machine, the interconnect bw could be more important than dram bw.

exclamaforte commented on torch/_inductor/analysis/device_info.py:
```diff
@@ -0,0 +1,136 @@
+# TODO investigate profiler support for tf32 and allow device to report correct number when it's turned on.
+_device_mapping: dict[str, DeviceInfo] = {
+    # Source: https://resources.nvidia.com/en-us-tensor-core/nvidia-tensor-core-gpu-datasheet
+    "NVIDIA H100": DeviceInfo(
+        tops={
+            torch.float64: 9.7,
+            torch.float32: 19.5,
+            torch.bfloat16: 1979.0,
+            torch.float16: 1979.0,
+            torch.float8_e8m0fnu: 3958.0,
+            torch.float8_e8m0fnu: 3958.0,
+            torch.float8_e4m3fnuz: 3958.0,
```
interested in Bert's opinion too, I would think that `current_clock_rate/default sufficient` would be fine considering that most of the flops calculation are just clockrate * core count * flops per core.

exclamaforte commented on torch/_inductor/analysis/profile_analysis.py:
```diff
@@ -0,0 +1,684 @@
+
+
+@register_adapter(["convolution", "_convolution", "cudnn_convolution"])
+def conv_adapter(
+    shapes: tuple[Any, ...], concrete: tuple[Any, ...]
+) -> tuple[tuple[Any], dict[Any, Any]]:
+    tmp = list(shapes)
+    if len(tmp) == 4:
+        transposed = False
+
+    transposed = bool(tmp[6])
+    tmp[6] = transposed
+
+    kwargs = {}
```
The json trace args is just the input args dimensions in a list (convenient). For convolution, if Transposed is set, then the flops counter also needs "out_val", which at runtime contains the output tensor. I don't think it's necessary, basically we could move the `conv_out_dims` function inside the flops counter, but I didn't want to mess with the flops counter in this PR. 

Other than that the len(tmp) == 4 comes from sometimes the input args only had 4 dims, not sure why. Maybe `convolution` vs `_convolution`...

In summary, I think we can do it from just the schema, but it requires the flops counter to be smarter. I expected other ops to have more conversions, but it's really just convolution.

eellison replied:
Couldn't we create FakeTensors with the input shapes, then run FlopCounter with them ?

exclamaforte replied:
So I just implemented this, but I think the code ends up looking worse because when I call the size function directly, it knows to ignore all of the other inputs like `SymInt groups`, `bool deterministic`, whereas if I run the op they get typechecked in c++

eellison replied:
I still think the current impl is wrong.. we should just instantiate the tensors, and then run FlopCounter mode, and run the op. you dont need to run the size function directly.

exclamaforte replied:
When you instantiate the tensors and run the ops, it goes through the dispatcher and all the inputs get typed/correctness checked, so I have to parse the string representations of all the kwargs in the profile event. I think it's worse than parsing only the shapes and creating fake tensors to pass in to the size functions.

exclamaforte commented on torch/_inductor/analysis/profile_analysis.py:
```diff
@@ -0,0 +1,684 @@
+    def _compute_stats(self) -> None:
+        """populates the name -> stats map"""
+        for event in self.events:
+            if "cat" not in event or "args" not in event or event["cat"] != "kernel":
```
we're not excluding cat, only testing cats that are "kernel"?

exclamaforte commented on torch/_inductor/utils.py:
```diff
@@ -2838,3 +2856,61 @@ def get_ld_library_path() -> str:
+    lines.append("|".join(f" {h:{w}} " for h, w in zip(headers, widths)))
+    #              widths          whitespace      horizontal separators
+    total_width = sum(widths) + (len(widths) * 2) + (len(widths) - 1)
+    lines.append("-" * total_width)
```
They're used in `profile_analysis.py`?

exclamaforte commented on torch/_inductor/analysis/profile_analysis.py:
```diff
@@ -0,0 +1,684 @@
+            dur = event["dur"]
+            if "kernel_flop" in event["args"]:
+                assert dur != 0
+                # 1000ms/s * flop / ms
```
added to backlog

eellison commented on torch/_inductor/analysis/device_info.py:
```diff
@@ -0,0 +1,150 @@
+    dram_gb: float
+
+
+# Indexing is based on `torch.cuda.get_device_name()`
+# TODO investigate profiler support for tf32 and allow device to report correct number when it's turned on.
+_device_mapping: dict[str, DeviceInfo] = {
```
Can we file an issue for this as a follow up ? It's not great that we are not doing this programatically..

exclamaforte replied:
done https://github.com/pytorch/pytorch/issues/153901

eellison commented on torch/_inductor/graph.py:
```diff
@@ -652,32 +653,24 @@ def is_small_channel(n: torch.fx.Node) -> bool:
-                        node_type = "in_out"
-                    else:
-                        node_type = "default"
+                counted_flops = count_flops_fx(node)
```
Can we make `count_flops_fx` a little bit smarter where it does not try to run this unless the node has a registration in flop_counter ? currently we always instantiate the tensors, and faketensormode, and run ?

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -426,12 +427,21 @@ def estimate_kernel_num_bytes(self):
+                if countable_fx(fx_node):
+                    f = count_flops_fx(fx_node)
+                    if f is not None:
+                        return V.graph.sizevars.size_hints((f,), fallback=0)[0]
```
nit: `size_hints(f,)[0] -> size_hint(f)`

eellison commented on torch/_inductor/utils.py:
```diff
@@ -3072,3 +3088,50 @@ def get_ld_library_path() -> str:
+    d2_default: Any = None,
+) -> Generator[tuple[Any, Any, Any], None, None]:
+    """
+    Zip two dictionaries together, indicating missing keys.
```
how does this indicate missing keys ?

eellison commented on torch/utils/flop_counter.py:
```diff
@@ -146,8 +145,8 @@ def conv_flop_count(
-@register_flop_formula([aten.convolution, aten._convolution])
-def conv_flop(x_shape, w_shape, _bias, _stride, _padding, _dilation, transposed, *args, out_shape=None, **kwargs) -> int:
+@register_flop_formula([aten.convolution, aten._convolution, aten.cudnn_convolution])
+def conv_flop(x_shape, w_shape, bias, stride, padding, dilation, transposed, *args, out_shape=None, **kwargs) -> int:
```
nit: reason why we needed to change arg names ?

eellison commented on torch/_inductor/analysis/profile_analysis.py:
```diff
@@ -0,0 +1,580 @@
+    shapes: tuple[Any, ...], concrete: tuple[Any, ...]
+) -> tuple[tuple[Any], dict[Any, Any]]:
+    tmp = list(shapes)
+    if len(tmp) == 4:
+        transposed = False
+
+    transposed = bool(tmp[6])
+    tmp[6] = transposed
```
If len(tmp) == 4, why are we indexing into tmp[6]?

eellison commented on torch/_inductor/analysis/profile_analysis.py:
```diff
@@ -0,0 +1,580 @@
+    return flop_function(*args, **kwargs)
+
+
+def _estimate_gb(event: dict[str, Any]) -> float:
```
When is this running? for the triton kernels, we should already have this, right ?

I mentioned this in a review internally - if we're going to estimate gb for mms, we should use the mm data access formula.

Generally - we have the gb from inductor right ? Also, inductor will dedupe inputs with same buffer, so i dont know that this comment is accurate.

exclamaforte replied:
> When is this running? for the triton kernels, we should already have this, right ?

For triton kernels, if they've been run with the correct configs they will have `kernel_num_gb` set, but this is supposed to augment a trace after it's been run even without those annotations.

> if we're going to estimate gb for mms, we should use the mm data access formula.

So the one on TritonTemplateKernel that gets logged in `inductor_meta` is just going to look at the input/output sizes and dedup if they're the same input/output buffers. I agree we should do better. I don't want to expand the scope too much, but something like flops_counter.py but for reads makes sense to me (probably as a followup.) Thoughts?

> Also, inductor will dedupe inputs with same buffer, so I don't know that this comment is accurate.

Moot point because the formula is bad anyways, but at the point of reading this profile, we've lost all the identifying information of the tensors and are left with just the shapes and dtypes so I don't think we can tell that they're the same buffer.

exclamaforte replied:
Updated the reads formulas, this code will be refactored into https://github.com/pytorch/pytorch/pull/154050
once it's ready.

eellison commented on torch/_inductor/analysis/profile_analysis.py:
```diff
@@ -0,0 +1,580 @@
+
+    flop_function = flop_registry[op_obj]
+
+    if "Input Dims" not in event["args"] or "Concrete Inputs" not in event["args"]:
+        breakpoint()
```
breakpoint() in code - is this being tested ?

exclamaforte replied:
It was functioning as an assert during debugging, I changed it to an assert.

eellison reviewed (commented):
Can we add a simple pointwise fusion and check that the bandwidth is somewhat in range we would expect ?

exclamaforte commented on torch/_inductor/utils.py:
```diff
@@ -3072,3 +3090,50 @@ def get_ld_library_path() -> str:
+
+
+def zip_dicts(
+    dict1: dict[Any, Any],
+    dict2: dict[Any, Any],
+    d1_default: Any = None,
+    d2_default: Any = None,
```
There's no type restriction here tho? Both the keys and values can be any type and they don't have to match up.

exclamaforte replied:
But there's no type restrictions so the variables don't make sense?

eellison replied:
with typevars, that is.

eellison replied:
(but it was a nit because it's up to you to address or not)

exclamaforte replied:
(fixed)

exclamaforte committed:
tests + profile analysis

exclamaforte committed:
all stuff

exclamaforte committed:
fix profile_analysis

exclamaforte committed:
fix tests

shoumikhin commented:
https://github.com/pytorch/pytorch/pull/154838 to resolve the url linter issues.
Feel free to add `skip-url-lint` label to this PR if needed

eellison commented on torch/_inductor/analysis/device_info.py:
```diff
@@ -0,0 +1,162 @@
+    If one is missing, please run DeviceInfo.get_device_info() and add it to _device_mapping.
+      name (str): name of the device to lookup. Should map onto torch.cuda.get_device_name().
+    """
+    if name not in _device_mapping:
+        return None
+    return _device_mapping[name]
```
nit: `return _device_mapping.get(name, None)`

eellison commented on torch/_inductor/analysis/device_info.py:
```diff
@@ -0,0 +1,162 @@
+        return None
+    device_info = lookup_device_info(name)
+    if device_info is None:
+        log_str = f"Device {name} not in datasheet, returning None"
+        log.info(log_str)
```
nit: `log.info(Device %s not in datasheet, returning None, name"` this will lazily format the strings, and not format if logging is not enabled.

eellison reviewed (approved):
🚢  

As an important follow up, can we make sure the mm kernels use the right memory access formula for gb accessed ? Arguably we should not include bandwidth % for any of the kernels that have a flop formula, for now, because it is misleading. 

cc @shunting314 to review as well

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
lines

malfet commented:
@pytorchbot revert -m "Broke rocm, see https://hud.pytorch.org/hud/pytorch/pytorch/642687af29718dbb75f6f38f4370a2f9bfd37348/1?per_page=50&name_filter=linux-jammy-rocm&mergeEphemeralLF=true" -c ignoredsignal

exclamaforte committed:
fix stuff

exclamaforte commented:
@pytorchbot merge

exclamaforte commented:
@pytorchbot merge

exclamaforte commented:
@pytorchbot merge

exclamaforte commented:
@pytorchbot rebase

exclamaforte commented:
@pytorchbot merge

exclamaforte commented:
@pytorchbot merge

exclamaforte commented:
@pytorchbot rebase

exclamaforte commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "broke a bunch of tests internally D76299454, probably also broke rocm inductor/test_analysis.py::TestAnalysisCUDA::test_augment_trace_against_flop_counter_maxat0_cuda_float16 [GH job link](https://github.com/pytorch/pytorch/actions/runs/15545277599/job/43766911025) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/060838c2312ad207c7afe2c86f8a484afea5f328)" -c nosignal

exclamaforte committed:
Inductor logging + analysis of torch.profile (#149697)

Summary:
Prereqs:
 - https://github.com/pytorch/pytorch/pull/152708

Features:
1. Adds inductor's estimate of flops and bandwidth to the json trace events that perfetto uses.
1. Only use the tflops estimation from triton if we don't have the info from the datasheet because Triton's estimates are inaccurate. I have a backlog item to fix triton flops estimation upstream. New `DeviceInfo` class, and new function `get_device_tflops`.
1. New helpers `countable_fx` and `count_flops_fx` helps get the flops of an `fx.Node`.
1. Extends Triton `torch.profiler` logging to `DebugAutotuner`.
1. New script `profile_analysis.py`: `--augment_trace` adds perf estimates to any perfetto json trace, `--analyze` creates a summary table of these perf estimates, and `--diff` will compare two traces side by side:
```python
Device(NVIDIA H100, 0):
 Kernel Name                              | resnet Kernel Count | resnet FLOPS       | resnet bw gbps        | resnet Dur (ms)    | resnet Achieved FLOPS % | resnet Achieved Bandwidth % | newresnet Kernel Count | newresnet FLOPS    | newresnet bw gbps     | newresnet Dur (ms) | newresnet Achieved FLOPS % | newresnet Achieved Bandwidth % 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 triton_poi_fused__native_batch_norm_legi | 24                  | 0                  | 0.11395268248131513   | 2.5919166666666666 | 0                       | 0.003401572611382541        | 24                     | 0                  | 0.11395268248131513   | 2.5919166666666666 | 0                          | 0.003401572611382541           
 sm90_xmma_fprop_implicit_gemm_f32f32_tf3 | 142                 | 16932673552.422373 | 0.2585007824198784    | 12.441619718309857 | 0.08683422334575583     | 0.007716441266265022        | 142                    | 16932673552.422373 | 0.2585007824198784    | 12.441619718309857 | 0.08683422334575583        | 0.007716441266265022           
 triton_red_fused__native_batch_norm_legi | 39                  | 0                  | 0.13990024992108846   | 5.752589743589743  | 0                       | 0.004176126863316074        | 39                     | 0                  | 0.13990024992108846   | 5.752589743589743  | 0                          | 0.004176126863316074           
 triton_poi_fused__native_batch_norm_legi | 25                  | 0                  | 0.31824055917536503   | 2.5291999999999994 | 0                       | 0.009499718184339253        | 25                     | 0                  | 0.31824055917536503   | 2.5291999999999994 | 0                          | 0.009499718184339253           
 void cutlass::Kernel2<cutlass_80_tensoro | 98                  | 16211056473.596165 | 0.42972434051025826   | 7.130408163265306  | 0.08313362294151874     | 0.012827592254037562        | 98                     | 16211056473.596165 | 0.42972434051025826   | 7.130408163265306  | 0.08313362294151874        | 0.012827592254037562           
 triton_red_fused__native_batch_norm_legi | 73                  | 0                  | 0.3225381327611705    | 9.987068493150682  | 0                       | 0.009628003963020014        | 73                     | 0                  | 0.3225381327611705    | 9.987068493150682  | 0                          | 0.009628003963020014           
 triton_poi_fused__native_batch_norm_legi | 15                  | 0                  | 1.4491211346487216    | 4.439333333333333  | 0                       | 0.043257347302946926        | 15                     | 0                  | 1.4491211346487216    | 4.439333333333333  | 0                          | 0.043257347302946926           
 void cutlass::Kernel2<cutlass_80_tensoro | 186                 | 14501701145.337954 | 0.2667131401910989    | 7.873865591397849  | 0.07436769818122027     | 0.007961586274361157        | 186                    | 14501701145.337954 | 0.2667131401910989    | 7.873865591397849  | 0.07436769818122027        | 0.007961586274361157           
 triton_poi_fused__native_batch_norm_legi | 33                  | 0                  | 1.4924556538193923    | 4.3101515151515155 | 0                       | 0.044550915039384846        | 33                     | 0                  | 1.4924556538193923    | 4.3101515151515155 | 0                          | 0.044550915039384846           
 triton_red_fused__native_batch_norm_legi | 29                  | 0                  | 0.25562590522631107   | 6.296275862068965  | 0                       | 0.007630624036606301        | 29                     | 0                  | 0.25562590522631107   | 6.296275862068965  | 0                          | 0.007630624036606301           
 triton_poi_fused__native_batch_norm_legi | 13                  | 0                  | 0.5870562174192726    | 2.7397692307692307 | 0                       | 0.01752406619162008         | 13                     | 0                  | 0.5870562174192726    | 2.7397692307692307 | 0                          | 0.01752406619162008            
 triton_poi_fused__native_batch_norm_legi | 34                  | 0                  | 0.41409928846284      | 2.853588235294117  | 0                       | 0.012361172789935523        | 34                     | 0                  | 0.41409928846284      | 2.853588235294117  | 0                          | 0.012361172789935523           
 triton_per_fused__native_batch_norm_legi | 34                  | 0                  | 0.11705315007018151   | 3.460647058823529  | 0                       | 0.0034941238826919864       | 34                     | 0                  | 0.11705315007018151   | 3.460647058823529  | 0                          | 0.0034941238826919864          
 triton_poi_fused__native_batch_norm_legi | 16                  | 0                  | 0.17207853197124584   | 2.3459375000000002 | 0                       | 0.005136672596156592        | 16                     | 0                  | 0.17207853197124584   | 2.3459375000000002 | 0                          | 0.005136672596156592           
 triton_per_fused__native_batch_norm_legi | 30                  | 0                  | 0.2639714322022256    | 6.131199999999999  | 0                       | 0.007879744244842555        | 30                     | 0                  | 0.2639714322022256    | 6.131199999999999  | 0                          | 0.007879744244842555           
 sm90_xmma_fprop_implicit_gemm_f32f32_tf3 | 100                 | 11875430356.891787 | 0.19494470869421385   | 16.36534           | 0.06089964285585531     | 0.005819245035648175        | 100                    | 11875430356.891787 | 0.19494470869421385   | 16.36534           | 0.06089964285585531        | 0.005819245035648175           
 triton_poi_fused__native_batch_norm_legi | 8                   | 0                  | 0.9854096626224687    | 3.2757500000000004 | 0                       | 0.029415213809625928        | 8                      | 0                  | 0.9854096626224687    | 3.2757500000000004 | 0                          | 0.029415213809625928           
 void cublasLt::splitKreduce_kernel<32, 1 | 56                  | 34377923395.147064 | 0.8310300045762317    | 3.4199999999999986 | 0.17629704305203628     | 0.024806865808245714        | 56                     | 34377923395.147064 | 0.8310300045762317    | 3.4199999999999986 | 0.17629704305203628        | 0.024806865808245714           
 triton_poi_fused__native_batch_norm_legi | 23                  | 0                  | 0.9944002965861103    | 3.2431304347826084 | 0                       | 0.02968359094286896         | 23                     | 0                  | 0.9944002965861103    | 3.2431304347826084 | 0                          | 0.02968359094286896            
 triton_per_fused__native_batch_norm_legi | 10                  | 0                  | 0.1826801058931057    | 4.428800000000001  | 0                       | 0.00545313748934644         | 10                     | 0                  | 0.1826801058931057    | 4.428800000000001  | 0                          | 0.00545313748934644            
 triton_poi_fused__native_batch_norm_legi | 10                  | 0                  | 0.3168973585366449    | 2.5471999999999997 | 0                       | 0.009459622642884923        | 10                     | 0                  | 0.3168973585366449    | 2.5471999999999997 | 0                          | 0.009459622642884923           
 triton_poi_fused__native_batch_norm_legi | 34                  | 0                  | 1.1463614897015777    | 4.124323529411764  | 0                       | 0.03421974596124114         | 34                     | 0                  | 1.1463614897015777    | 4.124323529411764  | 0                          | 0.03421974596124114            
 void cask_plugin_cudnn::xmma_cudnn::init | 44                  | 44045510816.64277  | 2.0661232850348643    | 3.6887499999999993 | 0.22587441444432194     | 0.06167532194133924         | 44                     | 44045510816.64277  | 2.0661232850348643    | 3.6887499999999993 | 0.22587441444432194        | 0.06167532194133924            
 sm90_xmma_fprop_implicit_gemm_f32f32_tf3 | 95                  | 7876855400.165316  | 0.4694941555946739    | 18.224315789473682 | 0.04039413025725802     | 0.014014750913273854        | 95                     | 7876855400.165316  | 0.4694941555946739    | 18.224315789473682 | 0.04039413025725802        | 0.014014750913273854           
 triton_per_fused__native_batch_norm_legi | 41                  | 0                  | 0.06825669875995298   | 3.0384146341463416 | 0                       | 0.002037513395819492        | 41                     | 0                  | 0.06825669875995298   | 3.0384146341463416 | 0                          | 0.002037513395819492           
 triton_poi_fused__native_batch_norm_legi | 23                  | 0                  | 0.08808154712430301   | 2.3275652173913044 | 0                       | 0.0026292999141582997       | 23                     | 0                  | 0.08808154712430301   | 2.3275652173913044 | 0                          | 0.0026292999141582997          
 triton_per_fused__native_batch_norm_legi | 40                  | 0                  | 0.18179321034952417   | 4.556825           | 0                       | 0.005426662995508183        | 40                     | 0                  | 0.18179321034952417   | 4.556825           | 0                          | 0.005426662995508183           
 triton_poi_fused__native_batch_norm_legi | 15                  | 0                  | 0.5887415155454232    | 2.783866666666667  | 0                       | 0.017574373598370836        | 15                     | 0                  | 0.5887415155454232    | 2.783866666666667  | 0                          | 0.017574373598370836           
 void cutlass::Kernel2<cutlass_80_tensoro | 38                  | 14242013806.264643 | 0.256592404353939     | 7.217631578947369  | 0.0730359682372546      | 0.007659474756834           | 38                     | 14242013806.264643 | 0.256592404353939     | 7.217631578947369  | 0.0730359682372546         | 0.007659474756834              
 triton_poi_fused__native_batch_norm_legi | 21                  | 0                  | 0.5842860973430516    | 2.7779047619047623 | 0                       | 0.017441376040091088        | 21                     | 0                  | 0.5842860973430516    | 2.7779047619047623 | 0                          | 0.017441376040091088           
 triton_per_fused__native_batch_norm_legi | 16                  | 0                  | 0.11509365173486417   | 3.5959375000000002 | 0                       | 0.0034356313950705724       | 16                     | 0                  | 0.11509365173486417   | 3.5959375000000002 | 0                          | 0.0034356313950705724          
 triton_poi_fused__native_batch_norm_legi | 14                  | 0                  | 0.1704672000243914    | 2.4044285714285714 | 0                       | 0.00508857313505646         | 14                     | 0                  | 0.1704672000243914    | 2.4044285714285714 | 0                          | 0.00508857313505646            
 triton_poi_fused__native_batch_norm_legi | 58                  | 0                  | 2.307520779930795     | 8.190706896551722  | 0                       | 0.06888121731136704         | 58                     | 0                  | 2.307520779930795     | 8.190706896551722  | 0                          | 0.06888121731136704            
 triton_per_fused__native_batch_norm_legi | 29                  | 0                  | 0.037243248971881276  | 3.0277586206896556 | 0                       | 0.001111738775280038        | 29                     | 0                  | 0.037243248971881276  | 3.0277586206896556 | 0                          | 0.001111738775280038           
 triton_poi_fused__native_batch_norm_legi | 20                  | 0                  | 0.04741699795428918   | 2.2911500000000005 | 0                       | 0.0014154327747549007       | 20                     | 0                  | 0.04741699795428918   | 2.2911500000000005 | 0                          | 0.0014154327747549007          
 triton_per_fused__native_batch_norm_legi | 25                  | 0                  | 0.13357016893727824   | 3.37536            | 0                       | 0.003987169222008305        | 25                     | 0                  | 0.13357016893727824   | 3.37536            | 0                          | 0.003987169222008305           
 triton_poi_fused__native_batch_norm_legi | 13                  | 0                  | 0.3089862268300253    | 2.8111538461538457 | 0                       | 0.009223469457612694        | 13                     | 0                  | 0.3089862268300253    | 2.8111538461538457 | 0                          | 0.009223469457612694           
 triton_poi_fused__native_batch_norm_legi | 17                  | 0                  | 0.3129385387909844    | 2.673              | 0                       | 0.009341448919133863        | 17                     | 0                  | 0.3129385387909844    | 2.673              | 0                          | 0.009341448919133863           
 triton_per_fused__native_batch_norm_legi | 19                  | 0                  | 0.2215568162533158    | 3.8837368421052636 | 0                       | 0.0066136363060691275       | 19                     | 0                  | 0.2215568162533158    | 3.8837368421052636 | 0                          | 0.0066136363060691275          
 std::enable_if<!(false), void>::type int | 23                  | 504916805.19297093 | 1.0118296096314707    | 8.113913043478261  | 0.0025893169497075447   | 0.030203868944223014        | 23                     | 504916805.19297093 | 1.0118296096314707    | 8.113913043478261  | 0.0025893169497075447      | 0.030203868944223014           
 triton_poi_fused_add_copy__38            | 56                  | 0                  | 0                     | 2.132482142857143  | 0                       | 0                           | 56                     | 0                  | 0                     | 2.132482142857143  | 0                          | 0                              
 triton_poi_fused_convolution_0           | 18                  | 0                  | 0.43458610794936897   | 2.773333333333334  | 0                       | 0.012972719640279667        | 18                     | 0                  | 0.43458610794936897   | 2.773333333333334  | 0                          | 0.012972719640279667           
 triton_poi_fused_convolution_1           | 17                  | 0                  | 0.028816312469162712  | 2.6145882352941174 | 0                       | 0.0008601884319153051       | 17                     | 0                  | 0.028816312469162712  | 2.6145882352941174 | 0                          | 0.0008601884319153051          
 void convolve_common_engine_float_NHWC<f | 44                  | 8641868995.31118   | 0.024730540008465626  | 25.87327272727273  | 0.04431727689903169     | 0.0007382250748795709       | 44                     | 8641868995.31118   | 0.024730540008465626  | 25.87327272727273  | 0.04431727689903169        | 0.0007382250748795709          
 triton_per_fused__native_batch_norm_legi | 12                  | 0                  | 0.6809930918986744    | 4.82675            | 0                       | 0.020328151996975356        | 12                     | 0                  | 0.6809930918986744    | 4.82675            | 0                          | 0.020328151996975356           
 triton_per_fused__native_batch_norm_legi | 14                  | 0                  | 0.02883030597936608   | 2.6651428571428575 | 0                       | 0.0008606061486377935       | 14                     | 0                  | 0.02883030597936608   | 2.6651428571428575 | 0                          | 0.0008606061486377935          
 triton_per_fused__native_batch_norm_legi | 16                  | 0                  | 0.0014658988233201874 | 2.098              | 0                       | 4.375817383045335e-05       | 16                     | 0                  | 0.0014658988233201874 | 2.098              | 0                          | 4.375817383045335e-05          
 triton_poi_fused__native_batch_norm_legi | 13                  | 0                  | 0.9926297180284697    | 3.2367692307692306 | 0                       | 0.02963073785159611         | 13                     | 0                  | 0.9926297180284697    | 3.2367692307692306 | 0                          | 0.02963073785159611            
 triton_poi_fused__native_batch_norm_legi | 9                   | 0                  | 1.3008817095666507    | 3.0863333333333336 | 0                       | 0.03883228983781048         | 9                      | 0                  | 1.3008817095666507    | 3.0863333333333336 | 0                          | 0.03883228983781048            
 void at::native::(anonymous namespace):: | 98                  | 0                  | 0.09174335613709389   | 4.408520408163265  | 0                       | 0.0027386076458833994       | 98                     | 0                  | 0.09174335613709389   | 4.408520408163265  | 0                          | 0.0027386076458833994          
 void at::native::vectorized_elementwise_ | 7                   | 0                  | 0                     | 1.7278571428571428 | 0                       | 0                           | 7                      | 0                  | 0                     | 1.7278571428571428 | 0                          | 0                              
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov


Differential Revision: D76435845

Pulled By: exclamaforte

exclamaforte committed:
some of the older gpus have issues collecting

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
remove failing tests

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
lint

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
lintrunner

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
remove tests that cause internal failures (merge later)

exclamaforte committed:
rename max_clockwork_mhz

exclamaforte committed:
skipifrocm

exclamaforte committed:
lint

exclamaforte committed:
fix error on AMD

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
fix asserts

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
profile analysis

exclamaforte committed:
stuff

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
remove inconsistent test

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
fix amd

exclamaforte commented:
@pytorchbot merge

exclamaforte committed:
lintrunner

exclamaforte commented:
@pytorchbot merge
